File: src/main/java/uk/edu/populationfitness/models/genes/bitset/NormalizingBitSetGenes.java
package uk.edu.populationfitness.models.genes.bitset;

import uk.edu.populationfitness.models.Config;

/**
 * Adds normalizing of fitness to caching and interpolation.
 */
public abstract class NormalizingBitSetGenes extends CachingInterpolatingBitSetGenes {

    private double normalizationRatio;

    private boolean isNormalisationRatioSet;

    protected NormalizingBitSetGenes(Config config, double maxInterpolatedValue) {
        super(config, maxInterpolatedValue);
        isNormalisationRatioSet = false;
    }

    /***
     * Implement this to define the normalization ratio
     * @param n
     * @return the normalization ratio
     */
    protected abstract double calculateNormalizationRatio(int n);

    @Override
    public double fitness() {
        if (isFitnessStored()) {
            return storedFitness();
        }

        if (!isNormalisationRatioSet){
            normalizationRatio = calculateNormalizationRatio(numberOfIntegers());
            isNormalisationRatioSet = true;
        }

        return storeFitness(calculateFitnessFromGenes(asDoubles()) / normalizationRatio);
    }
}

File: DotNet/PopulationFitness/PopulationFitness/Models/Genes/BitSet/NormalizingBitSetGenes.cs
namespace PopulationFitness.Models.Genes.BitSet
{

    /**
     * Adds normalizing of fitness to caching and interpolation.
     */
    public abstract class NormalizingBitSetGenes : CachingInterpolatingBitSetGenes
    {
        private double _normalizationRatio;
        private bool _isNormalisationRatioSet;

        protected NormalizingBitSetGenes(Config config, double maxInterpolatedValue) : base(config, maxInterpolatedValue)
        {
            _isNormalisationRatioSet = false;
        }

        /***
         * Implement this to define the normalization ratio
         * @param n
         * @return the normalization ratio
         */
        protected abstract double CalculateNormalizationRatio(int n);

        public override double Fitness
        {
            get
            {
                if (IsFitnessStored())
                {
                    return StoredFitness();
                }

                long[] integers = AsIntegers;

                if (!_isNormalisationRatioSet)
                {
                    _normalizationRatio = CalculateNormalizationRatio(integers.Length);
                    _isNormalisationRatioSet = true;
                }

                return StoreFitness(CalculateFitnessFromIntegers(integers) / _normalizationRatio);
            }
        }
    }
}
